### ISSUE TITLE
Suggestion: option to re-route DirectX calls to Vulkan using DXVK and VKD3D-proton dlls for each 3d app or game to get better compatibility or perormance

### ISSUE THREAD

**Issue by rovickti** [vxiiduu/VxKex#180]
[DXVK](https://github.com/doitsujin/dxvk/releases) re-implements DrectX 9,10,11 calls using Vulkan support in the GPU driver. 

Comes with x86/X64 dlls named: d3d10_1.dll, d3d10.dll, d2d11.dll, **dxgi.dll**, d3d9.dll, d3d10core.dll

[VKD3D](https://github.com/HansKristian-Work/vkd3d-proton) re-implements DirectX 12 on top of Vulkan. It's used in proton to support Windows games on Linux under Windows emulation, e.g. on Steam Deck.

Comes with x86/X64 dlls named: d3d12.dll (more recent Vulkan 1.3 versions have d3d12core.dll, but these are not supported by Windows 7)

These libraries are meant to be used by simply dropping the DLLs to the executable folder. They allow DirectX 12 support in Windows 7, and the possibility of [better](https://www.makeuseof.com/dxvk-windows-guide/), or maybe less glitchy, performance using older DirectX depending on the application and GPU/driver.

To get DirectX 12 support, including in [Windows 7](
https://www.reddit.com/r/windows7/comments/lvnz0j/ac_valhalla_runs_under_windows_7_with_vkd3d/), VKD3D can be used with DXVK by [copying the dxgi.dll](https://www.techpowerup.com/forums/threads/dx12-to-vulkan-mod-for-older-gpus-on-windows.315239/) from DXVK alongside VkD3D DLLs.

The latest Windows 7 drivers support Vulkan 1.2. The readme.md in the source files accompanying older releases gives the required Vulkan version. The latest version to not require Vulkan 1.3+ are: DXVK 1.10.x (latest: [1.10.3](https://github.com/doitsujin/dxvk/releases/tag/v1.10.3)) and VKD3D 2.8 and earlier ([2.8](https://github.com/HansKristian-Work/vkd3d-proton/releases/tag/v2.8). These require Vulkan 1.1. 

I don't know if anyone has done a backport of a more recent version to Vulkan 1.2, in another github project somewhere.

The previous generation of Windows 7 drivers support Vulkan 1.0 (NVIDIA 391.x for GTX 500 and earlier). I don't know if earlier versions of DXVK or VKD3D ever supported these.

----

The problem:

https://github.com/doitsujin/dxvk/wiki/Windows

> Some games fail to load all required DXVK DLLs even when they are correctly placed next to the game's executable. Applications may choose to only load DLLs from the system's System32 directory rather than the executable path, in which case the game will either crash or continue using the native D3D implementation. A common scenario that causes a crash is when the game loads dxgi.dll from System32, but then uses the d3d11.dll provided by DXVK.

> Similarly, some games will refuse to run with DXVK DLLs even if it could theoretically work. This commonly affects games with anti-cheat systems, such as Genshin Impact.

> Note: Make sure to use the 32-bit DLLs for 32-bit games. Windows applications will never load DLLs of the wrong architecture.

> Note: DO NOT replace Windows DLLs in System32 or SysWOW64 with DXVK's. This will break your Windows install.

----

Proposed solution:

* Have an option to redirect DirectX API calls to functions in the DXVK and VKD3D DLLs, instead of where ever the 3d application looks for DLLs. Including in system32 or elsewhere.

* The location for DLLs could be user configurable to allow bugfix releases, or to use a specific older version of DXVK or VKD3D with certain applications or games etc. It's possible to bundle these DLLs with VxKex as a shortcut to getting some DX12 games working, or as a performance boost to systems with older GPUs and drivers.

Benchmarks and demos could be used to quickly test everything is working under VxKex:
* DirectX 12: [Quake RTX demo]
(https://www.nvidia.com/en-gb/geforce/rtx/), [GFXbench 5 multi API](https://gfxbench.com/windows-download/gfxbench/) 3dmark (Speedway, needs 6gb VRAM), Passmark, etc. 

* DirectX 11 and earlier: Unigine [Suporposition](https://benchmark.unigine.com/superposition), NVIDIA [DirectX 11 demos](https://www.nvidia.com/en-gb/geforce/technologies/dx11/applications/). GFXbench 4/5

Note : using something like DXVK/VKD3D in multiplayer games may get people banned if the game's community hasn't explicitly got them whitelisted. Maybe a warning to that effect might help.

----

Background info for Microsoft driver models, GPU driver support being dropped, and Vulkan emulation and Windows 7

Windows 7 doesn't support DCH drivers. It makes it inconvenient for driver manufacturers to support older OSes like Windows 7. Microsoft wants people to move to later OSes as they make most of the revenue through bundling of the Windows store, they've offered free upgrades to Win 7, as well as retroactively software blocked Windows 7 from installing on [Kabylake/Ryzen](https://www.extremetech.com/computing/246049-microsoft-may-blocking-windows-7-8-1-updates-new-intel-kaby-lake-amd-ryzen-systems) and later processors after their attempt to [block Skylake](https://www.extremetech.com/computing/248069-unofficial-patch-unblocks-windows-7-8-1-updates-kaby-lake-ryzen) failed. In fact, Microsoft at one point apparently even made the [Windows store necessary to distribute the UI components of DCH drivers](https://www.reddit.com/r/StallmanWasRight/comments/ff4f1k/microsoft_store_required_for_full_functionality/). [Abuse](https://en.wikipedia.org/wiki/Criticism_of_Microsoft#Vendor_lock-in) of control over APIs, Standards, and Windows OS is a core component of Microsoft's [Embrace, Extend, Extinguish](https://en.wikipedia.org/wiki/Embrace,_extend,_and_extinguish) strategy, so it should come as no surprise. 

Of course, big games companies have a lot of sway and the possibility of moving to an open API like Vulkan/OpenGL, and [Microsoft even brought back](https://devblogs.microsoft.com/directx/world-of-warcraft-uses-directx-12-running-on-windows-7/) DirectX12 as a special implementation for World of Warcraft in 2019 instead of bringing DirectX 12 support right from the start!.  This was even with Microsoft about to turn down support for Windows 7 further - it goes to show their motivations.

NVIDIA dropped support for standard drivers in Windows 7 in October 2021 (except for the odd security related fix). e.g, [30xx, 20xx and 16xx series](https://www.nvidia.co.uk/download/driverResults.aspx/220251/en-uk) and 47x for [10xx through to 6xx series](https://www.nvidia.co.uk/download/driverResults.aspx/220283/en-uk). NVIDIA was producing 16xx GPUs until [November 2022](https://www.pcgamer.com/nvidia-reportedly-halts-production-of-rtx-2060-and-gtx-1660-gpus-nearly-four-years-after-release/), these cards are still under warranty but have no drivers in Windows 7. GTX 1060 was [Steam most common GPU](https://arstechnica.com/gaming/2022/12/after-more-than-5-years-nvidias-gtx-1060-is-no-longer-steams-most-used-gpu/) as recently May 2022, yet doesn't have updated drivers in Windows 7. AMD also followed in dropping driver support after NVIDIA did.

The most recent Vulkan supported by 47x drivers is [Vulkan 1.2 and DirectX 11](https://us.download.nvidia.com/Windows/474.89/474.89-win8-win7-release-notes.pdf) for GTX 6xx to RTX 30xx. GTX 4xx to GTX 6xx have [39x.x](https://us.download.nvidia.com/Windows/391.35/391.35-win10-win8-win7-desktop-release-notes.pdf) which support Vulkan 1.0. 

Vulkan 1.3 and DirectX 12 is not supported in Windows 7 at all by NVIDIA.

**Comment by jonm58** [comment]
~~dude,too long~~
https://github.com/Mantamaehne/VKD3Don7

**Comment by rovickti** [comment]
VKD3Don7 is DirectX12 on Vulkan 1.1 from VKD3D 2.6, last Vulkan 1.1 is [2.8](https://github.com/HansKristian-Work/vkd3d-proton/releases/tag/v2.8).

**Comment by Squall-Leonhart** [comment]
Pointless.

**Comment by TiberiumFusion** [comment]
12on7 exists. And I'd rather have d3d running as it should than layers upon layers of hacks.

**Comment by rovickti** [comment]
> TiberiumFusion: ..And I'd rather have d3d running as it should, instead of layers upon layers of hacks.

This doesn't interfere with existing D3D usage, at all.

It's an _option_ to re-route D3D 9-12 calls through user supplied (and maybe bundled) D3D to Vulkan DLLs.

This is a thing that VxKex does normally - VxKex works by re-routing calls to missing and changed API functions to it's own  DLLs. This should be a pretty quick feature to implement given how VxKex works.

Note people on Windows 10/11 [use DXVK for D3D 9-11 to Vulkan](https://www.makeuseof.com/dxvk-windows-guide/) where games allow it and it helps performance, because older 3d app/game code doesn't automatically get updated to use newer APIs - programs tend to stick to the release API unless it's a massively popular live service type game.

>12on7 exists.

Does it? There's no native support on Windows 7 for DirectX 7 - neither NVIDIA or AMD drivers. 

The only DirectX12 support implemented was for World of Warcraft (though Microsoft talked about adding maybe 1 or 2 other games)..

And that implementation was partial:

 https://www.anandtech.com/show/14078/microsoft-brings-dx12-to-windows-7
> Microsoft does say that DX12 will offer more features on Windows 10, which makes sense since the graphics stack was designed for it right from the start, but if you do play World of Warcraft on Windows 7, you’re going to get a free performance boost.

And completely bizarre...

> There’s many advantages to using this API over traditional DX11, especially for threading and draw calls. But, Microsoft made the decision long ago to only support DirectX 12 on Windows 10, with its WDDM 2.0 driver stack. 

> Today’s announcement is a pretty big surprise on a number of levels. If Microsoft had wanted to back-port DX12 to Windows 7, you would have thought they’d have done it before Windows 7 entered its long-term servicing state. As it is, even free security patches for Windows 7 are set to end on January 14, 2020, which is well under a year away, and the company is actively trying to migrate users to Windows 10 to avoid having a huge swath of machines sitting in an unpatched state. In fact, [they are about to add a pop-up notification to Windows 7 to let users know that they are running out of support very soon.](https://www.zdnet.com/article/microsoft-to-start-nagging-users-in-april-about-the-january-2020-windows-7-end-of-support-deadline/) So adding a big feature like DX12 now not only risks undermining their own efforts to migrate people away from Windows 7, but also adding a new feature well after Windows 7 entered long-term support. It’s just bizarre.

..unless you take into account that MS was just reacting to players refusing to leave Windows 7 and the risk of Blizzard switching to open APIs being too great if they weren't kept happy:

https://www.theverge.com/2019/3/13/18263422/microsoft-directx-12-windows-7-world-of-warcraft-support
> Microsoft has, in the past, used DirectX improvements as a way to push gamers to upgrade to the latest version of Windows. It’s clear this hasn’t worked for World of Warcraft players, but it’s still surprising to see Microsoft add new features to Windows 7 just as it’s starting to encourage everyone to leave the operating system behind.

So there's no chance of the World of Warcraft DirectX 12 implementation helping. 

As for DXVK and VK3D, the more recent versions need Vulkan 1.3. So the best that can be had in Windows 7 are the older versions.

Of course, it would be nice to be able to load DHC drivers in Windows 7 to get the latest D3D 12 support, but that might be beyond the reach of VxKex, or at least not in the medium term. VxKex [readme](https://github.com/vxiiduu/VxKex/blob/master/README.md) did say the initial focus was on apps rather than games so 3d graphics API issues may not be a priority. This is a quick solution, that also enables faster D3D 9-11 support which will not be coming as NVIDIA/AMD drivers will not be improved.

---

> Squall-Leonhart: Pointless

Maybe you misunderstand. See under "common issues" with loading dlls based on location/names, as different applications bechave differently. DXVK and VKD3D projects are focused on emulation under linux, so they don't tackle the D3D -> Vulkan on Windows properly.:

https://github.com/doitsujin/dxvk/wiki/Windows

**Comment by CommonLoon102** [comment]
> Maybe you misunderstand. See under "common issues" with loading dlls based on location/names, as different applications behave differently. DXVK and VKD3D projects are focused on emulation under linux, so they don't tackle the D3D -> Vulkan on Windows properly.:
> 
> https://github.com/doitsujin/dxvk/wiki/Windows

https://github.com/doitsujin/dxvk/wiki/Windows#windows-7--81

> Windows 7 / 8.1
> Any issue reported with an outdated versions of Windows will be ignored. Please do not use DXVK as a way to work around a game's minimum system requirements and expect things to work; if a game requires Windows 10, it probably does so for a reason.

Hahahahaha

**Comment by rovickti** [comment]
Ok, lol. I don't think commenters here are technical

https://github.com/doitsujin/dxvk/wiki/Windows#windows-7--81
> Windows 7 / 8.1 Any issue reported with an outdated versions of Windows will be ignored.

>> Hahahahaha

You do realise you are talking in the issue tracker of a project aimed at bringing Windows 8/10/11 support to Windows 7, right? Because, you know, Windows 7 is so much preferred, that there were more people on Windows 7 than Linux, at the point Microsoft finally managed to convince Chrome and Firefox to drop support, which just cause Windows 7 browser forks like Supoermium/Thorium and r3dfox that support Windows 7. And that Windows 7 is has slightly less people than all the Linux types if you go by browser market share.

https://github.com/vxiiduu/VxKex/blob/master/README.md
> VxKex is a set of API extensions for Windows 7 that allow some Windows 8, 8.1 and 10-exclusive applications to run on Windows 7.

Also check out [One core API](https://github.com/Skulltrail192/One-Core-API-Binaries) for Windows XP, and Windows [Vista Extended Kernel](https://win32subsystem.live/extended-kernel/).
: )

**Comment by jonm58** [comment]
I see **each 3d app**,this bad

I suggest changing the title to: New graphics API available through VxKex

**Comment by rovickti** [comment]
> I see each 3d app,this bad

I mean there could be a global option too. An option that people could switch based on the apps/games they are using in a session. But that's up the VxKex developer. It's not a given that Vulkan will help performance or glitchiness, or be bug free. The benefits depend on your particular GPU,driver and each application.

**Comment by Squall-Leonhart** [comment]
> The only DirectX12 support implemented was for World of Warcraft (though Microsoft talked about adding maybe 1 or 2 other games they never did it as far as I know)..

Wrong.

RDR2, Cyberpunk, and even the d912pxy for guildwars 2 made use of 12on7, its not a secret implementation, and supports up till DXR 1.0.  What it doesn't support is implicit barriers, resource contextual uploads, DxMath and video decoders, nor dxgi2+ api's.

**Comment by TiberiumFusion** [comment]
> Ok, lol. I don't think commenters here are technical

Bold assumption, Sherlock. Your love of arguing and writing walls of text doesn't mean you are above the other people here.
Also "Ok, lol".

Since you are the white knight here to deliver us all from our idiotic ways, why don't you write some code to make it happen? You clearly want it so much more badly than the rest of us. It would be a great way to prove how right you are and to graduate your github account from "less than 1 month old, 0 code, 0 repositories" to something that indicates you are a "technical" person. Otherwise, go troll elsewhere.

> microsoft bad microsoft dictator microsoft always makes things worse

We are all well aware how M$ is screwing us. We wouldn't be using VxKex otherwise. If you weren't so in love with hearing yourself talk, you'd realize that.

**Comment by jonm58** [comment]
> > hahahaja
> 
> > > Ok, lol. I don't think commenters here are technical
> 
> > > > TiberiumFusion: Bold assumption, Sherlock. Your love of arguing and writing walls of text doesn't mean you are above the other people here. Also "Ok, lol".
> 
> That was in response to a person going "hahahaha", at a emulation _for Linux_ not supporting older Windows, in an issue tracker about providing recent Windows 8,10,11 support on Windows 7. And the prior post. It wasn't in repsonse to your post which happened before and which I replied to. The prior post just said "pointless" adding absolutely nothing in the form of reasoning, while the DXVK wiki explicitly talked the issues in Windows when people just copied DLLs to the executable folder. A wiki article I explicitly quoted in the starting post, under a section called "the problem". I was trying to be as clear as possible, imagine what would have happened if I hadn't been as detailed. Under those circumstances it's reasonable to say "ok lol" and observe the commenters may not be technical since they seem to not follow the issue, before calmly explaining it to them. I mean if someone just goes hahahaha, then ok lol in the next post is a reasonable response. Neither of the posts appeared to be calm discussions of the situation, and they didn't say anything.
> 
> You seem to take the comment about not possibly being technical personally. It was not meant in that fashion, or at you. Of course, this is a subjective reaction on your part, because you were hardly critical of someone misunderstanding and going hahahaha, instead of communicating a point.
> 
> I did reply to your post, before this, if you have anything else to comment about it, feel free to quote and say something.
> 
> This should be a straight forward topic about re-routing some API calls, I'm not sure where the non-constructive tones or strong feelings on the topic are coming from.
> 

I thought you meant **each** commenters here

**Comment by CommonLoon102** [comment]
Hahaha was for this specific part:

> if a game requires Windows 10, it probably does so for a reason.

What is complete bullshit. Many games work on win 7 which require win 10. The only reason win 10 requirement is M$'s greed.

**Comment by rovickti** [comment]
Right, I focused on the 1st sentence, not that one, as..

> https://github.com/doitsujin/dxvk/wiki/Windows#windows-7--81
> Please do not use DXVK as a way to work around a game's minimum system requirements and expect things to work; if a game requires Windows 10, it probably does so for a reason.

..I just took this sentence to mean something _technically right_, but for _hardware_ system requirements (so misleading). People on marginal hardware for a game, relying on a DXVK FPS boost, may end up with an unsatisfactory experience, as the FPS boost may be inconsistent - e.g. only applicable under some rendering situations like outdoors vs indoors, where different API usage and bottlenecks may cause unsatisfactory experience. 

For example, AMD users on Windows 7 are missing the very belated 2022 [DX 11](https://videocardz.com/newz/amd-has-rebuilt-its-directx11-driver-from-the-ground-up-10-better-performance-on-average) improvements with 10%+ better FPS, as AMD dropped support in in [June 2021](https://www.amd.com/en/support/kb/faq/pa-120), and DXVK was even more popular purely for FPS boosts in Win 10/11 among AMD users.

> What is complete bullshit. 

But, yeah, I agree that just talking vaguely about _system_ requirements like that in that paragraph was likely rhetorical BS, to support DXVKs position of rejecting Windows 7 support. My guess as to why would be that possibly either as Linuxland developers didn't particularly care and wanted to avoid the overhead, or the project's main [contractor](https://www.gamingonlinux.com/2018/09/an-interview-with-the-developer-of-dxvk-part-of-what-makes-valves-steam-play-tick/) is Valve. And Valve had a position of dropping Windows 7, as likely their team of [economists](https://www.gamedeveloper.com/business/valve-economist-varoufakis-studying-game-economies-is-crucial) kept pointing out users were better monetised buying Win 10/11 exclusive games on later Windows, which led to Valve discontinuing the Steam GUI client when embeded Chromium dropped support, instead of supporting Chromium themselves - which would be easy for a company of that size, even without relying on patches in browser port projects. DXVK have also rejected things like support for integer scaling in Windows for high resolution monitors (e.g. NVIDIA implemented only for 16xx and later GPUs) to avoid unnecessary monitor/GPU upscaling artifacts, [saying](https://togithub.com/doitsujin/dxvk/issues/3699#issuecomment-1790853977) it's out of scope as the goal is to make things work correctly and run well on Linux only.
----

My apologies for any offense I may have caused to anyone. :).
